### PR TITLE
Support for SYCL's basic and nd-range kernels

### PR DESCRIPTION
**Pull request by Ellen-Wittingen**
This pull request adds support for SYCL's basic and nd-range kernels.

It supports the creation of new kernels with SYCL's `submit` method:
```
#include <sycl/sycl.hpp>

void main() {
  sycl::queue myQueue;

  sycl::event myEvent = myQueue.submit(
    [&](sycl::handler& cgh) {
      cgh.parallel_for(sycl::nd_range<2>(sycl::range<3>(6,4), sycl::range<2>(2,2)),
        /*@
          requires 1 + 2 == 3;
          ensures 3 * 5 == 15;
        */
        [=] (sycl::nd_item<2> it) {
          int a = 10;
        }
      );
    }
  );
  int a = 5;
  myEvent.wait();
}
```
Internally this block of code transformed into a class with a run method with a parallel block inside it, which holds the lambda function and contract written in the `parallel_for` parameters. Then `sycl::event myEvent` is assigned to an instance of the newly created class and is forked. This creates the following AST (after the language-specific rewrite pass):
```
...
class SYCL_EVENT_CLASS {
    requires 2 > 0;
    requires 6 >= 0;
    requires 6 % 2 == 0;
    requires 2 > 0;
    requires 4 >= 0;
    requires 4 % 2 == 0;
    requires [2 - 0][2 - 0][4 / 2 - 0][6 / 2 - 0](1 + 2 == 3);
    ensures [2 - 0][2 - 0][4 / 2 - 0][6 / 2 - 0](3 * 5 == 15);
    run {
        par SYCL_ND_RANGE_KERNEL(
            int GROUP_ID_0 = 0 .. 6 / 2,
            int GROUP_ID_1 = 0 .. 4 / 2,
            int LOCAL_ID_0 = 0 .. 2,
            int LOCAL_ID_1 = 0 .. 2
        )
        requires 1 + 2 == 3;
        ensures 3 * 5 == 15; {
            int a;
            a = 10;
        }
    }
}
void main() {
    ref myQueue;
    SYCL_EVENT_CLASS sycl_event_ref;
    sycl_event_ref = new SYCL_EVENT_CLASS();
    fork sycl_event_ref;
    int a;
    a = 5;
    join sycl_event_ref;
}
```
In SYCL, the host code does not wait on the submitted kernel to end before continuing. Therefore, the class containing the generated parellel block is forked. And when `myEvent.wait()` is called, which makes the host wait on the kernel code to end, the class instance is joined.

The methods of SYCL's `item` and `nd_item` which return the current (linear) thread id and range make use of the methods and their contracts inside the `sycl.hpp` header file.

Examples / tests for SYCL's kernels can be found in `examples/concepts/sycl/kernels/` and `test/main/vct/test/integration/examples/SYCLSpec.scala`.

Lambda functions are only allowed in SYCL methods, if they are used anywhere else in C++ VerCors will throw an error.

To check if the kernel ranges are good, pre-conditions are added above the run method to check them (for example, for nd_ranges the global range should be divisible by the local range).

The comments on the [previous pull request](https://github.com/utwente-fmt/vercors/pull/1052) have also been addressed. The implementation of namespaces has changed, as I realized I only needed to be able to resolve methods inside them: now in the parsing stage the methods nested inside one or more namespaces get the names of the namespaces added in front of their name. This means that in the resolution stage, the methods can just be found by checking for a matching name, instead of the nested search it was before.

### CODE REVIEW

File: src/parsers/antlr4/LangCPPParser.g4 [hunk 1]
```diff
@@ -273,16 +273,16 @@ assignmentOperator:
 	| OrAssign;
 
 expression:
-    assignmentExpression
-    | expression Comma assignmentExpression;
+    assignmentExpression;
+//    |  expression Comma assignmentExpression; // Do not match as then 'given' with multiple vars goes here, whilst it should not
```
[the remark this answers is not in the document]
Comment (Ellen-Wittingen) [reply src/parsers/antlr4/LangCPPParser.g4:277]: Do you want me to also fix this for `CParser.g4` in this PR? Or are you fixing it in separate PR for C?

[the remark this answers is not in the document]
Comment (pieter-bos) [reply src/parsers/antlr4/LangCPPParser.g4:277]: Yes, please do.

File: src/rewrite/vct/rewrite/lang/LangCPPToCol.scala [hunk 2]
```diff
@@ -3,48 +3,226 @@ package vct.col.rewrite.lang
...
+    override def text: String = decl.o.messageInContext(s"Non-SYCL code inside command group declarations is not supported.")
+  }
+
+  private case class IncorrectParallelForLambdaArgument(o: Origin) extends UserError {
+    override def code: String = "incorrectParallelForLambdaArgument"
+    override def text: String = o.messageInContext("The second parameter for the parallel_for method, which is a lambda method, " +
+      "only takes one (nd_)item argument, which should match the first (nd_)range parameter of the parallel_for method.")
+  }
+
+  private case class WrongKernelDimensionType(expr: Expr[_]) extends VerificationFailure {
```
[the remark this answers is not in the document]
Comment (Ellen-Wittingen) [reply src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:84]: I had them as `VerificationFailure` as  then I could throw them using `BlameUnreachable`. These errors can only happen is there is a mistake in the rewrite implementation, the user should not be able to cause these, nor the verification backend. That is why I thought I should throw a `BlameUnreachable`, and thus would need them to be `VerificationFailure`s.

But if `VerificationFailure` and  `BlameUnreachable` are only for errors from the backend, then I can throw them as `UserError`s instead. Unless there is a specif error class for implementation error that I missed?

File: src/rewrite/vct/rewrite/lang/LangSpecificToCol.scala [hunk 3]
```diff
@@ -110,10 +109,11 @@ case class LangSpecificToCol[Pre <: Generation]() extends Rewriter[Pre] with Laz
     case JavaLocalDeclarationStatement(locals: JavaLocalDeclaration[Pre]) => java.initLocal(locals)
 
     case CDeclarationStatement(decl) => c.rewriteLocal(decl)
-    case CPPDeclarationStatement(decl) => cpp.rewriteLocal(decl)
+    case CPPDeclarationStatement(decl) => cpp.rewriteLocalDecl(decl)
     case goto: CGoto[Pre] => c.rewriteGoto(goto)
     case barrier: GpgpuBarrier[Pre] => c.gpuBarrier(barrier)
-
+    case eval@Eval(CPPInvocation(_, _, _, _)) => cpp.invocationStatement(eval)
+    case eval@Eval(PreAssignExpression(CPPLocal(_, _), _)) => cpp.preAssignStatement(eval)
```
[the remark this answers is not in the document]
Comment (Ellen-Wittingen) [reply src/rewrite/vct/rewrite/lang/LangSpecificToCol.scala:116]: `Eval(CPPInvocation(_, _, _, _))` only handles SYCL method calls that need to be rewritten to a statement. All other `CPPInvocation` that can be rewritten to an expression should be handled in the `Expr` dispatch method. So no error should be thrown there.

`Eval(PreAssignExpression(CPPLocal(_, _), _))` I will move to directly match the expression in the `Expr` dispatch method instead, as it is only used to throw an error if a SYCL event type is reassigned. In an older version of the code I needed to return a statement, which is why it ended up here. I must have missed this when checking my code before opening the PR.

File: src/parsers/vct/parsers/transform/CPPToCol.scala [hunk 4]
```diff
@@ -989,11 +1095,11 @@ case class CPPToCol[G](override val originProvider: OriginProvider, override val
       withModifiers(modifiers, mods =>
         Seq(new Predicate(args.map(convert(_)).getOrElse(Nil), convert(definition),
           mods.consume(mods.threadLocal), mods.consume(mods.inline))
-        (SourceNameOrigin(convert(name), origin(decl)))))
+        (SourceNameOrigin((currentNamespacePath.reverse :+ convert(name)).mkString("::"), origin(decl)))))
```
[the remark this answers is not in the document]
Comment (Ellen-Wittingen) [reply src/parsers/vct/parsers/transform/CPPToCol.scala:1098]: I indeed did not have time to implement namespace resolving fully, and using prepended strings was the fastest solution that still had all functionality I needed for the SYCL implementation

File: src/rewrite/vct/rewrite/lang/LangCPPToCol.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ellen-Wittingen) [reply src/rewrite/vct/rewrite/lang/LangCPPToCol.scala]: Just found the `Unreachable` class in `VerificationError.scala`. I will use that one instead for these kinds of errors

**Review by pieter-bos — APPROVED** [review]

**Pull request merged**